### PR TITLE
[WordFilter] Optimize find and replace

### PR DESCRIPTION
**Pull request by IAmTheRedSpy**
### Type

- [ ] Bugfix
- [x] Enhancement
- [ ] New feature

### Description of the changes
Change wordFilter's implementation so that it finds and replaces words more quickly.
The expected output is the same.

### CODE REVIEW

File: cogs/wordfilter/wordfilter.py [hunk 1]
```diff
@@ -403,14 +402,13 @@ def __init__(self, bot: Red):
...
-                self.logger.error("Exception!")
-                self.logger.error(error)
-                self.logger.info("Word: %s", word)
-                self.logger.info("Filtered message: %s", filteredMsg)
+        try:
+            filteredMsg = _filterWord(filteredWords, filteredMsg)
+        except Exception as error: # pylint: disable=broad-except
+            self.logger.error("Exception!")
+            self.logger.error(error)
+            self.logger.info("Word: %s", word)
```
Comment (Injabie3) [inline cogs/wordfilter/wordfilter.py]: This `word` variable no longer exists, since you removed the loop.

**Review by Injabie3 — CHANGES_REQUESTED** [review]
Thanks taiyaki, some things below.

I'm also running into problems with the list of filters (same as what you sent me in #bot-test).  Do we need to escape things (regex is not my strong point)?
```
Traceback (most recent call last):
  File "/home/ren/github/TestServer/TestServer/cogs/wordfilter/wordfilter.py", line 406, in checkWords
    filteredMsg = _filterWord(filteredWords, filteredMsg)
  File "/home/ren/github/TestServer/TestServer/cogs/wordfilter/wordfilter.py", line 476, in _filterWord
    return re.sub(regex, _censorMatch, string, flags=re.IGNORECASE)
  File "/home/ren/github/TestServer/lib/python3.7/re.py", line 192, in sub
    return _compile(pattern, flags).sub(repl, string, count)
  File "/home/ren/github/TestServer/lib/python3.7/re.py", line 286, in _compile
    p = sre_compile.compile(pattern, flags)
  File "/home/ren/github/TestServer/lib/python3.7/sre_compile.py", line 764, in compile
    p = sre_parse.parse(p, flags)
  File "/home/ren/github/TestServer/lib/python3.7/sre_parse.py", line 930, in parse
    p = _parse_sub(source, pattern, flags & SRE_FLAG_VERBOSE, 0)
  File "/home/ren/github/TestServer/lib/python3.7/sre_parse.py", line 426, in _parse_sub
    not nested and not items))
  File "/home/ren/github/TestServer/lib/python3.7/sre_parse.py", line 816, in _parse
    p = _parse_sub(source, state, sub_verbose, nested + 1)
  File "/home/ren/github/TestServer/lib/python3.7/sre_parse.py", line 426, in _parse_sub
    not nested and not items))
  File "/home/ren/github/TestServer/lib/python3.7/sre_parse.py", line 654, in _parse
    source.tell() - here + len(this))
re.error: multiple repeat at position 2661 (line 2, column 2176)
```

```
glory hole|bl[o0]*{0,3}w j[o0]*{0,3}bs
                  ^col 2176
```

File: cogs/wordfilter/wordfilter.py (hunk 1 above)
Comment (IAmTheRedSpy) [reply cogs/wordfilter/wordfilter.py]: That's from before I started editing it. Both `*` and `{}` are repetition for get rid of one.

**Comment by IAmTheRedSpy** [comment]
> Thanks taiyaki, some things below.
> 
> I'm also running into problems with the list of filters (same as what you sent me in #bot-test). Do we need to escape things (regex is not my strong point)?
> 
> ```
> Traceback (most recent call last):
>   File "/home/ren/github/TestServer/TestServer/cogs/wordfilter/wordfilter.py", line 406, in checkWords
>     filteredMsg = _filterWord(filteredWords, filteredMsg)
>   File "/home/ren/github/TestServer/TestServer/cogs/wordfilter/wordfilter.py", line 476, in _filterWord
>     return re.sub(regex, _censorMatch, string, flags=re.IGNORECASE)
>   File "/home/ren/github/TestServer/lib/python3.7/re.py", line 192, in sub
>     return _compile(pattern, flags).sub(repl, string, count)
>   File "/home/ren/github/TestServer/lib/python3.7/re.py", line 286, in _compile
>     p = sre_compile.compile(pattern, flags)
>   File "/home/ren/github/TestServer/lib/python3.7/sre_compile.py", line 764, in compile
>     p = sre_parse.parse(p, flags)
>   File "/home/ren/github/TestServer/lib/python3.7/sre_parse.py", line 930, in parse
>     p = _parse_sub(source, pattern, flags & SRE_FLAG_VERBOSE, 0)
>   File "/home/ren/github/TestServer/lib/python3.7/sre_parse.py", line 426, in _parse_sub
>     not nested and not items))
>   File "/home/ren/github/TestServer/lib/python3.7/sre_parse.py", line 816, in _parse
>     p = _parse_sub(source, state, sub_verbose, nested + 1)
>   File "/home/ren/github/TestServer/lib/python3.7/sre_parse.py", line 426, in _parse_sub
>     not nested and not items))
>   File "/home/ren/github/TestServer/lib/python3.7/sre_parse.py", line 654, in _parse
>     source.tell() - here + len(this))
> re.error: multiple repeat at position 2661 (line 2, column 2176)
> ```
> 
> ```
> glory hole|bl[o0]*{0,3}w j[o0]*{0,3}bs
>                   ^col 2176
> ```

That's from before I started editing it. Both `*` and `{}` are repetition so get rid of one of them.

File: cogs/wordfilter/wordfilter.py (hunk 1 above)
Comment (IAmTheRedSpy) [reply cogs/wordfilter/wordfilter.py]: It looks like `re.sub` can only except if the regex is invalid. So maybe we can just remove that exception?

Comment (Injabie3) [reply cogs/wordfilter/wordfilter.py]: Sure, let's take out this line, and change this except block to be more specific and catch `re.error` instead.

**Review by Injabie3 — APPROVED** [review]
Other than that from before, LGTM

**Pull request merged**